crates/cheatnet/src/execution/cheatable_syscall_handler.rs

```diff
@@ -0,0 +1,297 @@
+        reference_ids: &HashMap<String, usize>,
+        references: &[HintReference],
+    ) -> Result<Box<dyn Any>, VirtualMachineError> {
+        self.syscall_handler
+            .compile_hint(hint_code, ap_tracking_data, reference_ids, references)
+    }
+}
+
+// crates/blockifier/src/execution/syscalls/hint_processor.rs:454
+/// Retrieves a [Relocatable] from the VM given a [`ResOperand`].
+/// A [`ResOperand`] represents a CASM result expression, and is deserialized with the hint.
+fn get_ptr_from_res_operand_unchecked(vm: &mut VirtualMachine, res: &ResOperand) -> Relocatable {
+    let (cell, base_offset) = match res {
+        ResOperand::Deref(cell) => (cell, Felt252::from(0)),
+        ResOperand::BinOp(BinOpOperand {
+            op: Operation::Add,
+            a,
+            b: DerefOrImmediate::Immediate(b),
+        }) => (a, Felt252::from(b.clone().value)),
+        _ => panic!("Illegal argument for a buffer."),
+    };
+    let base = match cell.register {
+        Register::AP => vm.get_ap(),
+        Register::FP => vm.get_fp(),
+    };
+    let cell_reloc = (base + i32::from(cell.offset)).unwrap();
+    (vm.get_relocatable(cell_reloc).unwrap() + &base_offset).unwrap()
+}
+
+pub fn stark_felt_from_ptr_immutable(
+    vm: &VirtualMachine,
+    ptr: &Relocatable,
+) -> Result<StarkFelt, VirtualMachineError> {
+    Ok(felt_to_stark_felt(&felt_from_ptr_immutable(vm, ptr)?))
+}
+
+pub fn felt_from_ptr_immutable(
+    vm: &VirtualMachine,
+    ptr: &Relocatable,
+) -> Result<Felt252, VirtualMachineError> {
+    let felt = vm.get_integer(*ptr)?.into_owned();
+    Ok(felt)
+}
+
+fn get_syscall_operand(hint: &StarknetHint) -> Result<&ResOperand, HintError> {
+    let StarknetHint::SystemCall { system: syscall } = hint else {
+        return Err(HintError::CustomHint(
+            "Test functions are unsupported on starknet.".into(),
+        ));
+    };
+    Ok(syscall)
+}
+
+fn get_syscall_cost(syscall_selector: SyscallSelector) -> u64 {
+    match syscall_selector {
+        SyscallSelector::LibraryCall | SyscallSelector::CallContract => {
+            constants::CALL_CONTRACT_GAS_COST
+        }
+        SyscallSelector::Deploy => constants::DEPLOY_GAS_COST,
+        SyscallSelector::GetExecutionInfo => constants::GET_EXECUTION_INFO_GAS_COST,
+        _ => unreachable!("Syscall has no associated cost"),
+    }
+}
+
+impl CheatableSyscallHandler<'_> {
+    fn execute_next_syscall_cheated(
+        &mut self,
+        vm: &mut VirtualMachine,
+        hint: &StarknetHint,
+    ) -> HintExecutionResult {
+        // We peek into the selector without incrementing the pointer as it is done later
+        let syscall = get_syscall_operand(hint)?;
+        let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+        let selector =
+            SyscallSelector::try_from(stark_felt_from_ptr_immutable(vm, &initial_syscall_ptr)?)?;
+        self.verify_syscall_ptr(initial_syscall_ptr)?;
+
+        match selector {
```

Comment:
We could generalize it and provide two abstractions:
- override syscall implementation
- hook to execute after/before syscall

Can be left for future work.

Reply:
It's a good idea for #916